server/api/controllers/device.controller.js

```diff
@@ -91,6 +92,124 @@ module.exports = function DeviceController(gladys) {
+    });
+
+    // fix period
+    const beginDate = new Date();
+    switch (params.chartPeriod) {
+      case 'last2day-selector':
+        beginDate.setDate(beginDate.getDate() - 2);
+        break;
+      case 'last1week-selector':
+        beginDate.setDate(beginDate.getDate() - 7);
+        break;
+      case 'last1month-selector':
+        beginDate.setDate(beginDate.getDate() - 31);
+        break;
+      case 'last1year-selector':
+        beginDate.setDate(beginDate.getDate() - 365);
+        break;
+      case 'last1day-selector':
+      default:
+        beginDate.setDate(beginDate.getDate() - 1);
+    }
+
+    params.begin_date = beginDate;
+    params.end_date = new Date();
+
+    // Choose attributes
+    params.attributes_device = [];
+    params.attributes_device.push('name');
+    params.attributes_device.push('selector');
+    params.attributes_device_feature = [];
+    params.attributes_device_feature.push('id');
+    params.attributes_device_feature.push('name');
+    params.attributes_device_feature.push('selector');
+    params.attributes_device_feature.push('unit');
+    params.attributes_device_feature.push('last_value');
+    params.attributes_device_feature.push('last_value_changed');
+    params.attributes_device_feature.push('last_downsampling');
+    params.attributes_device_room = [];
+    params.attributes_device_room.push('name');
+    params.attributes_device_room.push('selector');
+
+    const devices = await gladys.device.getFeatureStates(params);
+
+    // Downsample result to reduce nb of value in response
+    if (params.downsample && params.downsample === 'true') {
+      if (devices && devices.length > 0) {
+        let chartWidth = 100;
+        if (params.maxValue) {
+          chartWidth = params.maxValue;
+        }
+        const featureArray = [];
+        const featureValuesArray = [];
+        devices.forEach((device) => {
+          device.features.forEach((feature) => {
+            if (feature.device_feature_states && feature.device_feature_states.length > 0) {
+              const newFeatureStateArray = [];
+              feature.device_feature_states.forEach(function changeState(state, index) {
+                newFeatureStateArray.push({
+                  x: new Date(state.created_at).getTime(),
+                  y: state.value,
+                });
+                featureValuesArray.push(state.value);
+              });
+
+              let smoothFeatureStates = newFeatureStateArray;
+              if (newFeatureStateArray.length > chartWidth) {
+                smoothFeatureStates = LTTB(newFeatureStateArray, chartWidth);
+              }
+              feature.device_feature_states = smoothFeatureStates;
+              feature.trend = newFeatureStateArray[newFeatureStateArray.length - 1] - newFeatureStateArray[0];
+
+              featureArray.push(feature);
+            }
+          });
+          device.features = featureArray;
```

Comment:
It's weird to do that in the controller. The controller should be as slim as possible!

It should be done in the model

Reply:
I consider downsampling after getting from database must be done in another part of medel access

Do you want put it in function device.getFeatureStates.js ?